File: pay-later-go/db/transaction.go
package db

import (
	"errors"

	"gorm.io/gorm"
)

// Transaction is a gorm database model represent the transactions data
type Transaction struct {
	gorm.Model
	UserID     uint
	MerchantID uint
	Amount     float64

	User     User
	Merchant Merchant
}

// CreditLimit is the error message when rejecting a transaction when
// processing the current transaction will result in his total credit amount
// to exceed his credit limit
const CreditLimit = "credit limit"

// CreateTransaction inserts a transaction record in the database
func (db *DB) CreateTransaction(
	user User,
	merchant Merchant,
	amount float64,
) (Transaction, error) {
	userTotalCredit := db.UserDues(user)
	if userTotalCredit+amount > user.CreditLimit {
		return Transaction{}, errors.New(CreditLimit)
	}

	txn := Transaction{
		UserID:     user.ID,
		MerchantID: merchant.ID,
		Amount:     amount,
	}
	result := db.db.Create(&txn)
	return txn, result.Error
}

// CreatePayback inserts a transaction record in the database as payback for
// a user. This does not correspond to a particular merchant since it's received
// by our service and so the merchant_id for this transaction will be 0.
func (db *DB) CreatePayback(
	user User,
	amount float64,
) (Transaction, error) {
	txn := Transaction{
		UserID: user.ID,
		Amount: -1 * amount,
	}
	result := db.db.Create(&txn)
	return txn, result.Error
}

File: pay-later-rs/src/db/transaction.rs
use anyhow::{anyhow, Result};
use sqlx::{Pool, Sqlite, FromRow};
use crate::db::{User, Merchant, Report};

#[derive(Debug, FromRow)]
pub struct Transaction {
    pub id: Option<i64>,
    pub user_id: i64,
    pub merchant_id: Option<i64>,
    pub amount: f64,
}

impl Transaction {
    #[allow(dead_code)]
    pub fn new(user_id: i64, merchant_id: Option<i64>, amount: f64) -> Self {
        Self { id: None, user_id, merchant_id, amount }
    }

    pub async fn create(user: User, merchant: Merchant, amount: f64, pool: &Pool<Sqlite>) -> Result<()> {
        let user_dues = Report::user_dues(user.name, pool).await?;
        if user_dues+amount > user.credit_limit {
            return Err(anyhow!("credit limit"));
        }

        sqlx::query("INSERT INTO transactions (user_id, merchant_id, amount) VALUES ($1, $2, $3)")
            .bind(user.id)
            .bind(merchant.id)
            .bind(amount)
            .execute(pool)
            .await?;
        Ok(())
    }

    pub async fn payback(user: User, amount: f64, pool: &Pool<Sqlite>) -> Result<()> {
        sqlx::query("INSERT INTO transactions (user_id, amount) VALUES ($1, $2)")
            .bind(user.id)
            .bind(-1.0 * amount)
            .execute(pool)
            .await?;
        Ok(())
    }
}
